presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -271,4 +278,75 @@ public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, N
                     .collect(toImmutableList());
         }
     }
+
+    private static class RemoveIdentityCastContext
+    {
+        private final Map<Symbol, Expression> expressionAssignments;
+        private final Map<Symbol, Type> typeAssignments;
+
+        public RemoveIdentityCastContext(Map<Symbol, Expression> expressionAssignments,
+                                         Map<Symbol, Type> typeAssignments)
+        {
+            this.expressionAssignments = expressionAssignments;
+            this.typeAssignments = typeAssignments;
+        }
+
+        public Map<Symbol, Expression> getExpressionAssignments()
+        {
+            return expressionAssignments;
+        }
+
+        public Map<Symbol, Type> getTypeAssignments()
+        {
+            return typeAssignments;
+        }
+    }
+
+    private static class RemoveIdentityCastsRewriter
+            extends ExpressionRewriter<RemoveIdentityCastContext>
+    {
+        @Override
+        public Expression rewriteExpression(Expression node, RemoveIdentityCastContext context,
+                                            ExpressionTreeRewriter<RemoveIdentityCastContext> treeRewriter)
+        {
+            if (context.getExpressionAssignments() == null || context.getTypeAssignments() == null) {
+                return node;
+            }
+            else {
+                Map<Symbol, Type> typeAssignments = context.getTypeAssignments();
+                for (Map.Entry<Symbol, Expression> expressionAssignment : context.getExpressionAssignments().entrySet()) {
```

Comment:
Why you have to iterate over all assignments here?
https://github.com/Teradata/presto/pull/329/commits/17a508110880b64c069957c8585b7d5f5653a291#diff-f5a0289b0bcc438ad2c235a5d431f773R116 will call `simplifyExpressions` on each of the entries from the `expressionAssignments` map.

Shouldn't this rewrite be limited to single expression only and be called for each expression independently (e.g. by `Maps.transformValues`)?